File: go/internal/agent/fixflow/analyze.go
package fixflow

import (
	"context"
	"fmt"
	"iter"
	"sort"
	"strings"

	"google.golang.org/adk/agent"
	"google.golang.org/adk/agent/workflowagents/parallelagent"
	"google.golang.org/adk/session"

	"automation-agent/internal/agent/setup"
)

const (
	editPrefix = "edit:" // state key per file: edit:<workPath> -> new content
	pathPrefix = "path:" // state key per file: path:<workPath> -> target edit path
)

// EditFunc produces the edit for one file's work: the target path (which may differ
// from the source path — e.g. a test file) and the new content. Return a zero
// FileEdit (empty Path or Content) to skip this file.
type EditFunc func(ctx context.Context, w FileWork) (FileEdit, error)

// ParallelAnalyze fans out one analyzer agent per FileWork (ADK parallel agents,
// each writing distinct state keys so they never collide), calls fn for each, and
// returns the collected non-empty edits sorted by path.
func ParallelAnalyze(ctx context.Context, work []FileWork, fn EditFunc) ([]FileEdit, error) {
	if len(work) == 0 {
		return nil, fmt.Errorf("analyze: no files to work on")
	}
	sorted := append([]FileWork(nil), work...)
	sort.Slice(sorted, func(i, j int) bool { return sorted[i].Path < sorted[j].Path })

	agents := make([]agent.Agent, 0, len(sorted))
	seen := make(map[string]int, len(sorted))
	for _, w := range sorted {
		a, err := newAnalyzer(uniqueAnalyzerName(seen, w.Path), w, fn)
		if err != nil {
			return nil, fmt.Errorf("build analyzer for %s: %w", w.Path, err)
		}
		agents = append(agents, a)
	}
	par, err := parallelagent.New(parallelagent.Config{AgentConfig: agent.Config{
		Name: "analyze_all", Description: "Per-file analysis in parallel", SubAgents: agents,
	}})
	if err != nil {
		return nil, fmt.Errorf("build analyzers: %w", err)
	}
	r, err := setup.NewRunner("fix-analyze", par)
	if err != nil {
		return nil, err
	}
	state, err := setup.DriveCollectState(ctx, r, "system", "analyze", "Produce the edits.")
	if err != nil {
		return nil, err
	}

	edits := make([]FileEdit, 0, len(sorted))
	for _, w := range sorted {
		content, _ := state[editPrefix+w.Path].(string)
		path, _ := state[pathPrefix+w.Path].(string)
		if strings.TrimSpace(content) != "" && path != "" {
			edits = append(edits, FileEdit{Path: path, Content: content})
		}
	}
	if len(edits) == 0 {
		return nil, fmt.Errorf("analyze produced no edits")
	}
	return edits, nil
}

// uniqueAnalyzerName derives a unique ADK sub-agent name from a path. SafeName maps every
// non-alphanumeric character to '_', so distinct paths (e.g. "a/b.kt" and "a-b.kt") can
// collapse to the same name; ParallelAgent requires unique sub-agent names, so a collision
// gets a numeric suffix — otherwise one analyzer silently shadows another and that file's
// edits are dropped. State keys use the full path, so they never collide.
func uniqueAnalyzerName(seen map[string]int, path string) string {
	base := "analyze_" + setup.SafeName(path)
	seen[base]++
	if n := seen[base]; n > 1 {
		return fmt.Sprintf("%s_%d", base, n)
	}
	return base
}

func newAnalyzer(name string, w FileWork, fn EditFunc) (agent.Agent, error) {
	return agent.New(agent.Config{
		Name:        name,
		Description: "Analyzes " + w.Path,
		Run: func(ctx agent.InvocationContext) iter.Seq2[*session.Event, error] {
			return func(yield func(*session.Event, error) bool) {
				edit, err := fn(ctx, w)
				if err != nil {
					yield(nil, fmt.Errorf("analyze %s: %w", w.Path, err))
					return
				}
				if edit.Path == "" || strings.TrimSpace(edit.Content) == "" {
					yield(setup.TextEvent(name, "skipped "+w.Path, nil), nil)
					return
				}
				yield(setup.TextEvent(name, "edited "+edit.Path, map[string]any{
					editPrefix + w.Path: edit.Content,
					pathPrefix + w.Path: edit.Path,
				}), nil)
			}
		},
	})
}

File: kotlin/src/main/kotlin/com/automation/agent/agent/fixflow/Analyze.kt
package com.automation.agent.agent.fixflow

import com.google.adk.kt.agents.BaseAgent
import com.google.adk.kt.agents.InvocationContext
import com.google.adk.kt.agents.ParallelAgent
import com.google.adk.kt.events.Event
import com.automation.agent.agent.setup.driveCollectState
import com.automation.agent.agent.setup.newRunner
import com.automation.agent.agent.setup.safeName
import com.automation.agent.agent.setup.textEvent
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow

private const val EDIT_PREFIX = "edit:" // state key per file: edit:<workPath> -> new content
private const val PATH_PREFIX = "path:" // state key per file: path:<workPath> -> target edit path

/**
 * Produces the edit for one file's work: the target path (which may differ from the source path —
 * e.g. a test file) and the new content. Return a zero [FileEdit] (empty path or content) to skip.
 */
fun interface EditFunc {
    suspend operator fun invoke(work: FileWork): FileEdit
}

/**
 * Fans out one analyzer agent per [FileWork] (ADK parallel agents, each writing distinct state keys
 * so they never collide), calls [fn] for each, and returns the collected non-empty edits sorted by
 * path.
 */
suspend fun parallelAnalyze(work: List<FileWork>, fn: EditFunc): List<FileEdit> {
    require(work.isNotEmpty()) { "analyze: no files to work on" }
    val sorted = work.sortedBy { it.path }

    val seen = mutableMapOf<String, Int>()
    val analyzers = sorted.map { AnalyzerAgent(uniqueAnalyzerName(seen, it.path), it, fn) }
    val parallel = ParallelAgent(name = "analyze_all", description = "Per-file analysis in parallel", subAgents = analyzers)
    val state = driveCollectState(newRunner("fix-analyze", parallel), "system", "analyze", "Produce the edits.")

    val edits =
        sorted.mapNotNull { w ->
            val content = state[EDIT_PREFIX + w.path] as? String
            val path = state[PATH_PREFIX + w.path] as? String
            if (!content.isNullOrBlank() && !path.isNullOrEmpty()) FileEdit(path = path, content = content) else null
        }
    require(edits.isNotEmpty()) { "analyze produced no edits" }
    return edits
}

/**
 * Derive a unique sub-agent name from a path. safeName maps every non-alphanumeric char to
 * `_`, so distinct paths (e.g. `a/b.kt` and `a-b.kt`) can collapse to the same name;
 * ParallelAgent needs unique sub-agent names, so a collision gets a numeric suffix —
 * otherwise one analyzer silently shadows another and that file's edits are dropped.
 */
private fun uniqueAnalyzerName(seen: MutableMap<String, Int>, path: String): String {
    val base = "analyze_${safeName(path)}"
    val n = (seen[base] ?: 0) + 1
    seen[base] = n
    return if (n > 1) "${base}_$n" else base
}

private class AnalyzerAgent(name: String, private val work: FileWork, private val fn: EditFunc) :
    BaseAgent(name = name, description = "Analyzes ${work.path}") {
    override fun runAsyncImpl(context: InvocationContext): Flow<Event> = flow {
        val edit = fn(work)
        if (edit.path.isEmpty() || edit.content.isBlank()) {
            emit(textEvent(name, "skipped ${work.path}"))
        } else {
            emit(textEvent(name, "edited ${edit.path}", mapOf(EDIT_PREFIX + work.path to edit.content, PATH_PREFIX + work.path to edit.path)))
        }
    }
}
